### ISSUE TITLE
OpenShot on Windows requires undocumented network access

### ISSUE THREAD

**Issue by russellvt** [OpenShot/openshot-qt#2305]
**Describe the bug**
Following installation, the Windows installer complains about requiring network access from launch.exe. Efforts to find documentation supporting this requirement have (so far) proven fruitless.

**System Details:**
 - Windows 7
 - OpenShot Version v2.4.3 for Windows

**To Reproduce**
1. Install OpenShot on Windows 7 from https://www.openshot.org/
2. Once setup is complete, choose "Launch"
3. Observe the Windows Security Alert from "C:\program files\openshot video editor\launch.exe"
4. Notice the "Publisher" for the Windows Security Alert is "Unknown" (read: Unsigned)
5. See screenshot, below.

**Expected behavior**
Network access requirements should be documented, or contain some sort of warning. Windows binaries should be signed by a verifiable publisher.

Current web searches only show typical responses for "trojan'd" software (generally from other scanner output listing OpenShot in the Windows Firewall whitelist, ironically enough).

**Screenshots**
![image](https://user-images.githubusercontent.com/915922/47692085-d68c1880-dbb1-11e8-9e14-7cb70789d6ab.png)

**Logs**
N/A

**Additional context**
Hopefully, there's at least a general "README" to be had for non-UNIX users. No apparent requirement in launch.py for network, though that's only a top-level script for python modules, obviously.

It would likely be a good exercise to launch the Linux version and observe the network requirements (which I have yet to do). Unfortunately, the low-level troubleshooting on Windows 7 (and others) often leaves much to be desired.

Assuming this *might* have something related to "Automatically send errors and metrics to OpenShot" (which defaults as "Yes, I would like to improve OpenShot!")... though obviously this action is taken *prior* to actual approval from the user.

**Comment by peanutbutterandcrackers** [comment]
It is for the error metrics collection (google analytic). Open-shot does not need any other network use (except, maybe check for updates; OpenShot notifies users of available updates when a new stable version is out, so perhaps that).

Yes, we could have done things a little better. But we are short on active volunteers. But hopefully this will get addressed in the future releases. Thank you for notifying us.

**Comment by DylanC** [comment]
@peanutbutterandcrackers @ferdnyc - I feel maybe we should add this to documentation but where is the question? The GitHub readme.md?

**Comment by ferdnyc** [comment]
Doesn't Windows have an installer? IMHO, a disclosure about data collection should be prominently featured (not buried-in-a-long-rambling-wall-of-legalese type "featured") during the steps taken to install OpenShot, so that the user has been informed _before_ it's even fully available to run for the first time.

Bonus points if data collection could be opted out of right there, but just giving precise instructions on _how_ to opt out after starting OpenShot for the first time would suffice.

I'd also personally like to see the option to disable data collection moved from the "Debug" tab of the Preferences to a more appropriate one. "General" would be fine.

And in addition, yeah, both the Github project README and the [openshot.org Download page](https://openshot.org/download/) should (again IMHO) make much more prominent disclosures regarding data collection.

It's really impossible to make _too many_ mentions of this kind of thing. People just automatically ignore future repetitions after they've seen it, anyway.

**Comment by ferdnyc** [comment]
(Pre-launch opt-out would involve the installer either writing a `%USERPROFILE%/.openshot_qt/openshot.settings` that's customized from the standard `_default.settings` by having `send_metrics` already set to `false`, or writing a separate `%USERPROFILE%/.openshot_qt/no-send-metrics` or whatever that OpenShot looks for when creating a new preferences file with `_default.settings`.)

**Comment by peanutbutterandcrackers** [comment]
@DylanC - for now we could probably start with the FAQ section int he README, perhaps?

**Comment by ferdnyc** [comment]
> (Pre-launch opt-out would involve the installer either writing a `%USERPROFILE%/.openshot_qt/openshot.settings` that's customized from the standard `_default.settings` by having `send_metrics` already set to `false`, or writing a separate `%USERPROFILE%/.openshot_qt/no-send-metrics` or whatever that OpenShot looks for when creating a new preferences file with `_default.settings`.)

Actually, come to think of it this is wrong, since opting out on install should disable collection for _all_ users of OpenShot on that system. not just the one running the installer. So, the right thing to do would be to modify the `_default.settings` that gets installed into `C:\Program Files\OpenShot Video Editor\settings`, changing `send_metrics` from `true` to `false`.

**Comment by teknowledgist** [comment]
Opt-in for tracking would be preferred, but I'd be OK with having an install switch that allows turning it off automatically (e.g. for deployment to a classroom).  If this is possible, documentation is sorely missing.

**Comment by teknowledgist** [comment]
In case it's useful for someone else, here is some PowerShell code to disable the metrics by default (and disable the tutorial by default):

````
$text = Get-Content -Path 'C:\Program Files\OpenShot Video Editor\settings\_default.settings' -ReadCount -raw
$newtext = $text -replace ',\r\n',',' -split '\r\n' | ForEach-Object {
               if (($_ -match 'send_metrics') -or ($_ -match 'tutorial_enabled')) {
                  $_ -replace '("value":\s*)true','$1false'
               } else {$_}
            }
[IO.File]::WriteAllLines('C:\Program Files\OpenShot Video Editor\settings\_default.settings', ($newtext.replace(',',",`r`n")))
````
It will not turn off metrics for users who have already started OpenShot, but all new users will have data collection disabled.

**Comment by thomasaarholt** [comment]
I just arrived here after wondering a lot why the Windows installer included a checkbox for "add exception to Windows firewall". 

However, upon first open, I see dialog that OpenShot defaults to not sending metrics, with the option to turn that on. I approve of this :) (this change wasn't mentioned in the discussion above, so I am mentioning it here for the worried future user)

![image](https://user-images.githubusercontent.com/2721423/83941606-01f2d280-a7ed-11ea-9164-c48a96d023fa.png)

**Comment by demisjohn** [comment]
I also came here wondering why OpenShot had a cryptic "Windows Firewall exception" on the installer.

It looks like the "Would you like to improve OS?" dialog should be shown during the Install, and the firewall opened only if user says Yes.
Or If the user later chooses "Yes!" then ask for admin PW and unblock it.

Currently this is in the wrong order - It asks users whether to unblock FW without any explanation (checked by default - hopefully all users uncheck it?), THEN later asks if you'd like to contribute via network access.

It made me wonder if OS would not function without network access, since it's in the installer.

**Comment by dscotese** [comment]
If any of you have questions, please ask.  If you don't, please consider my question: Why didn't you do what I just did?  It took me about 5 minutes.  Well, 5 minutes plus about 5 years of experience messing around in github.  Hence my invitation to answer questions.  I like empowering people.

**Comment by demisjohn** [comment]
Thanks @dscotese for applying a change.  Not sure how long that took you to do, but if I had tried to do the same, I guarantee it would have taken me ~2 days to figure it out and applying it in a way without damaging something else.  Your expertise is very much appreciated.

**Comment by dscotese** [comment]
I'm getting old, so I like to pass my expertise on.  Here's the basic outline:

1. Use Github's search function to find "Firewall" in "this repository only" since that string must be provided somewhere (this method has failed me before, I guess because github didn't search the ".json" files in the other repo, but maybe it was some other problem).
1. Examine the code to see if the string "Add an exception to the Windows Firewall" is easily altered (Yay! It was).
1. Click the little pencil (and let GIthub make me a fork of the project so I can edit "my own version").
1. Use the drop-down at the top to select a branch other than "master," (in my "dscotese:OpenShot" fork) or create a new branch.
1. Save the file, and then (?) follow the instructions to create a pull request.
1. Receive tips and help from other devs and apply them or ask for more help.
1. Further edits to the files in the same branch are _automatically_ included in the pull request.
1. WIN!

**Comment by ferdnyc** [comment]
On @demisjohn 's point:

> I also came here wondering why OpenShot had a cryptic "Windows Firewall exception" on the installer.
> 
> It looks like the "Would you like to improve OS?" dialog should be shown during the Install, and the firewall opened only if user says Yes.
> Or If the user later chooses "Yes!" then ask for admin PW and unblock it.
> 
> Currently this is in the wrong order - It asks users whether to unblock FW without any explanation (checked by default - hopefully all users uncheck it?), THEN later asks if you'd like to contribute via network access.

It **used** to be that OpenShot's anonymized telemetry was opt-out, and during the first startup would be active _until_ you opted out, meaning that some data would be sent back on initial launch. I agree that was the wrong order (and argued mightily against it), and the code's since been changed so that nothing is sent until after the first tutorial popup with the checkbox has been shown.

But one of the consequences of the previous wrong ordering was, on Windows the **first** thing users would see when they launched OpenShot after installing it was a Windows Firewall "network activity detected" dialog — not exactly friendly, and even less explanation as to why that'd be needed. So the firewall rule addition was included in the installer to prevent that.

Creating the firewall rule in the Windows installer makes sense, as it's already being run with Administrator permissions. Creating the rule after launch would require a second UAC prompt (which some people would argue is correct and necessary, but others would view as an unnecessary annoyance), and more importantly would require Windows-specific code in our cross-platform Python application, which is something we try to avoid whenever possible.

> It made me wonder if OS would not function without network access, since it's in the installer.

It would, and no functionality would be affected except _perhaps_ the version update checks. We wouldn't get anonymized stats and crash reports, but that's perfectly fine.

**Comment by VA1DER** [comment]
It is three years later after this bug was opened and OpenShot's installer is still asking to add a wide open firewall hole with no explanation.  And if you say no, it asks again when it launches.  And this for something that is, ostensibly, not even opted into by default.

First of all, good job for making the errors and metrics an opt-in setting.  +1.  But negative several million points for the firewall hole being a) there in the first place, b) opt out, and c) for asking twice (at installation and at first run).  The stated explanation of "don't have enough volunteers" doesn't pass the smell test.  Why is this issue closed?  It's still doing it as of 2.5.1.

The reasons stated here for the firewall rule doesn't match what it's asking for.  It is asking for an INCOMING firewall rule.  How is that remotely required for metrics and error reporting?

**Comment by dscotese** [comment]
@VA1DER, It makes sense to me that "sending anonymized usage and error information" might require both incoming and outgoing exceptions, but I have not examined the code. If you know better, please update the explanation as I did [here](https://github.com/OpenShot/openshot-qt/issues/2305#issuecomment-699709243) and we can make OpenShot even better.  Thanks!

**Comment by VA1DER** [comment]
A firewall exception is not required for any program to initiate an internet connection with the outside world.  Your web browser didn't need a firewall exception.  If the program on your computer is initiating the connection, no firewall exception required.  Once it initiates the connection, data can flow both ways.

The fact that OpenShot needs an incoming connection firewall exception means that something in OpenShot wants to listen for connections initiated FROM the outside.  What possible reason could OpenShot have for that?  This is unnecessary and a huge security risk.

I'll tell you right now, there is no error reporting or usage information sending which should require that.  There is NO valid reason that anyone has given for OpenShot to require this.  Full stop.

**Comment by marcello-pietrobon** [comment]
The same with me. I subscribe everything VA1DER says.
Unless you want remote debugging there's no need for it, and even in such case the user should be asked to open a firewall only when needed.
Worse than that, if I block the access with an application firewall, then the application crashes on the start.
Disappointed. Full stop for me too.

**Comment by LostPr3acherProject** [comment]
Passive observer comment here: I have paid close attention to this whole thread, both to the critique and to the developer response. Very "hinky" response, I am going to have to say. I worked with cyber security on both Windows and Linux in the military, and I can tell you there are zero reasons I can think of for a video editing program to need INCOMING access through a firewall - only outgoing. Very uncool to blame it on lack of staff, and less cool to gaslight people for asking legitimate questions about the issue. The snark in the responses is rather telling. As a user, I'm unsure what to think about this one. Perhaps the developer was having a bad day and hadn't had coffee when they chose to reply. Either way, a little more graciousness in the reply format from said developer would avoid a lot of vinegar in the tea, so to speak. I hope this gets resolved soon.

**Comment by dscotese** [comment]
Really, we are all potentially developers.  Github doesn't call anyone on this thread a developer, just "contributor" and "collaborator." I like to take the heat myself, so I'll assume that "If any of you have questions, please ask. If you don't, please consider my question: Why didn't you do what I just did?" is the "snarky" part.  I suppose that for some people, being shown doors they didn't perceive previously, and asked why they haven't opened them or at least knocked can be interpreted as snarky.  I won't apologize for that or stop doing it, because it is the kind of opportunity that I've been given over and over by people willing to be "snarky" to me, which I took, and from which I learned a great deal ("plus about 5 years of experience messing around in github.").

One of the most common ways of getting problems solved is to make the people who eventually solve them feel like crap until they actually solve them.  I don't think it's a very good strategy.  I much prefer to figure out the best way *I* might start (or continue) the solution, engage in that, and ask questions so I can learn to get better at it and encourage others to do the same.  That has always been my strategy on github, and I recommend it going forward.  I opened this comment thread today in the hope that I might find a question I could answer to help someone contribute to the project.  Speaking up is a good first step, because you open the door to learning more.  @LostPr3acherProject 's experience with security on Windows and Linux is probably quite valuable to the thread and regarding this issue.  What we really need is someone with the time to verify it works perfectly well when the incoming access is denied, delve into the code to remove the request for it, and submit a pull request.  I'll have that time some day, but I don't currently.

**Comment by Kennna** [comment]
I just installed this program for the first time and was quite surprised that it didn't open at all on my computer (which is set up to block all new program's internet access by default, until turned on manually by me).

Sounds like we're dealing with a trojan horse in disguise here.

**Comment by USATechDude** [comment]
I've verified with several customers that our installer does not require Internet access while installing the program.  After installation, the default settings prevent OpenShot from accessing the Internet. You have to opt-in before OpenShot starts sending metrics across the Internet (using Google Analytics to gather anonymous data).  I've worked with a few school districts that have stringent Internet access requirements.  OpenShot meets their needs since OpenShot doesn't collect personal information and it doesn't transmit information across the Internet without user authorization.

Please ensure that you download and install the official versions of our software from [our website](https://www.openshot.org/download/) or our [GitHub Repository](https://github.com/OpenShot/openshot-qt/releases/tag/v2.5.1).

Also, this Issue is closed.  If you want to involve our developers, please [open a new issue](https://github.com/OpenShot/openshot-qt/issues/new/choose).  They will not see your comments added to the issue.  I only stumbled across this issue by happenstance.

**Comment by Kennna** [comment]
Correct, the installation doesn't require internet access. But you cannot start the software itself without granting access.

**Comment by USATechDude** [comment]
I don't understand why my systems work differently than yours, and I would like to help you resolve the issue.  If you would like to work one-on-one, you can send me an E-mail at support@openshot.org.

**Comment by muchqs** [comment]
this software triggered my firewall after installation when attempting to run the program itself. When loopback access is denied by firewall this program automatically terminates itself and refuses to run. Consistent behavior with what Kennna described above.
Uninstalled.

**Comment by dscotese** [comment]
It's possible that the binaries that are getting downloaded are different from the binaries you'd get if you compiled the code on your own.  That might explain @USATechDude's systems working differently (from @Kennna's) , but there is another explanation...

@Kennna explained that without Internet access, OS won't start.  @USATechDude said that __installing__ it doesn't require access, and he may not have noticed the difference, even though @Kennna made it explicit, and @muchqs provided some more detail.  I suggested before that someone could fix this on their own and make a pull request, but it sounds like openshot.org (who own this repo) would not merge it, so if anyone does, they can advertise that they have a version of OS that doesn't use the Internet at all.
